### PR TITLE
[3689] Preserve channel list state

### PR DESCRIPTION
**Pull request by bychkovdmitry**
### 🎯 Goal

**Important: this PR only addresses the issue on the channel list screen.**

Restore scroll state on the channel list screen.

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/9600921/176408372-cbd9192a-22af-4e74-acb4-e71d68dee0ea.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/9600921/176406447-aedf544b-eb65-43e2-ae0a-b0df41343371.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

**Option 1:**

1. Navigate to the channel list screen 
2. Scroll down
3. Open a channel
4. Press back

**Option 2:**

1. Navigate to the channel list screen 
2. Scroll down
3. Rotate the screen

Expected: the previous channel list scroll position is restored.

Also, we are now exposing channel list internals so that clients are able to control scroll state themselves. Apply the patch below to see how clients can now scroll to a specific item programmatically:

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/list/ChannelListFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/list/ChannelListFragment.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/list/ChannelListFragment.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/list/ChannelListFragment.kt	(revision b37cc8b87f621bd6a2ecbf7125d5ed5696e9a55f)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/channel/list/ChannelListFragment.kt	(date 1656506040949)
@@ -129,6 +129,14 @@
                     .findNavController(R.id.hostFragmentContainer)
                     .navigateSafely(direction)
             }
+
+            // Scroll the list to the 15th position once the dataset is committed
+            setChannelListUpdateListener {
+                val position = 15
+                if (it.size > position) {
+                    getLayoutManager()?.scrollToPosition(position)
+                }
+            }
         }
 
         binding.searchInputView.apply {

```

</details>

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/common/internal/SnapToTopDataObserver.kt [hunk 1]
```diff
@@ -51,6 +51,8 @@ internal class SnapToTopDataObserver(
             return
         }
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/common/internal/SnapToTopDataObserver.kt:53]: This is also relevant to the fix as previously it tried to scroll to the top even if there were no previous items and `layoutManager.findFirstVisibleItemPosition()` returned `-1`.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/ChannelListView.kt [hunk 2]
```diff
@@ -114,6 +139,51 @@ public class ChannelListView : FrameLayout {
         configureDefaultMoreOptionsListener(context)
     }
 
+    /**
+     * Returns the inner [RecyclerView] that is used to display a list of channel list items.
+     *
+     * @return The inner [RecyclerView] with channels.
+     */
+    public fun getRecyclerView(): RecyclerView {
```
Comment (bychkovdmitry) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/channel/list/ChannelListView.kt:147]: The naming is taken from `MessageListView` as we already expose the inner RecyclerView there with  `MessageListView.getRecyclerView()`

**Review by filbabic — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Review by filbabic — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Bugs validated (bugfixes)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Pull request merged**